lambdas/services/bulk_upload_report_service.py

```diff
@@ -16,31 +17,188 @@ class BulkUploadReportService:
+            ods_reports: list[OdsReport] = self.generate_ods_reports(
+                report_data, generated_at
             )
-            self.write_items_to_csv(report_data, f"/tmp/{file_name}")
+            logger.info("Successfully processed daily ODS reports")
+
+            self.generate_summary_report(ods_reports, generated_at)
+            logger.info("Successfully processed daily summary report")
+
+        else:
+            logger.info("No data found, no new report file to upload")
+
+    def generate_ods_reports(
+        self, report_data: list[dict], generated_at: str
+    ) -> list[OdsReport]:
+        ods_reports: list[OdsReport] = []
 
-            logger.info("Uploading new report file to S3")
+        grouped_ods_data = {}
+        for item in report_data:
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
 
-            self.s3_service.upload_file(
-                s3_bucket_name=staging_bucket_name,
-                file_key=f"reports/{file_name}",
-                file_name=f"/tmp/{file_name}",
+            if uploader_ods_code not in grouped_ods_data:
+                grouped_ods_data[uploader_ods_code] = []
+            grouped_ods_data[uploader_ods_code].append(item)
+
+        for ods_code, ods_data in grouped_ods_data.items():
+            ods_report = self.generate_individual_ods_report(
+                ods_code, grouped_ods_data[ods_code], generated_at
             )
+            ods_reports.append(ods_report)
+
+        return ods_reports
+
+    def generate_individual_ods_report(
+        self, ods_code: str, ods_report_data: list[dict], generated_at: str
+    ) -> OdsReport:
+        total_successful = set()
+        total_registered_elsewhere = set()
+        total_suspended = set()
+        failures_per_patient = {}
+        unique_failures = {}
+
+        logger.info(f"Generating ODS report file for {ods_code}")
+
+        for item in ods_report_data:
+            nhs_number = item.get(MetadataReport.NhsNumber)
+            upload_status = item.get(MetadataReport.UploadStatus)
+            pds_ods_code = item.get(MetadataReport.PdsOdsCode)
+            uploader_ods_code = item.get(MetadataReport.UploaderOdsCode)
+
+            if upload_status == "complete":
+                total_successful.add(nhs_number)
+
+                if uploader_ods_code != pds_ods_code:
+                    total_registered_elsewhere.add(nhs_number)
+
+            if upload_status == "suspended":
+                total_suspended.add(nhs_number)
+
+            failure_reason = item.get(MetadataReport.FailureReason, "")
+            if failure_reason:
+                if nhs_number not in unique_failures:
+                    unique_failures.update({nhs_number: failure_reason})
+
+        for reason in unique_failures.values():
+            if reason not in failures_per_patient:
+                failures_per_patient[reason] = 0
+            failures_per_patient[reason] += 1
```

Comment:
`for reason in unique_failures.values():
    failures_per_patient[reason] = failures_per_patient.get(reason, 0) + 1`